qi-lib/flow/core/deforest-syntax.rkt

```diff
@@ -0,0 +1,293 @@
+                  (#%host-expression init)
+                  _)))))
+
+(define-syntax-class cad*r-datum
+  #:attributes (countdown)
+  (pattern (~datum car) #:attr countdown #'0)
+  (pattern (~datum cadr) #:attr countdown #'1)
+  (pattern (~datum caddr) #:attr countdown #'2)
+  (pattern (~datum cadddr) #:attr countdown #'3)
+  (pattern (~datum caddddr) #:attr countdown #'4)
+  (pattern (~datum cadddddr) #:attr countdown #'5))
+
+(define-syntax-class fsc-list-ref
+  #:attributes (pos name)
+  #:datum-literals (#%host-expression #%blanket-template __ _ #%fine-template list-ref)
+  (pattern (~or (#%fine-template
+                 ((#%host-expression list-ref) _ idx))
+                (#%blanket-template
+                 ((#%host-expression list-ref) __ idx)))
+           #:attr pos #'idx
+           #:attr name #'list-ref)
+  (pattern (~or (esc (#%host-expression cad*r:cad*r-datum))
+                (#%fine-template
+                 ((#%host-expression cad*r:cad*r-datum) _))
+                (#%blanket-template
+                 ((#%host-expression cad*r:cad*r-datum) __)))
+           #:attr pos #'cad*r.countdown
+           #:attr name #'cad*r))
+
+(define-syntax-class fsc-length
+  #:datum-literals (#%host-expression #%blanket-template __ _ #%fine-template length)
+  (pattern (~or (esc
+                 (#%host-expression length))
+                (#%fine-template
+                 ((#%host-expression length) _))
+                (#%blanket-template
+                 ((#%host-expression length) __)))))
+
+(define-syntax-class fsc-empty?
+  #:datum-literals (#%host-expression #%blanket-template __ _ #%fine-template empty? null?)
+  (pattern (~or (esc
+                 (#%host-expression (~or empty?
+                                         null?)))
+                (#%fine-template
+                 ((#%host-expression (~or empty?
+                                          null?)) _))
+                (#%blanket-template
+                 ((#%host-expression (~or empty?
+                                          null?)) __)))))
+
+(define-syntax-class fsc-default
+  #:datum-literals (#%host-expression #%blanket-template __ _ #%fine-template cstream->list)
+  (pattern cstream->list))
+
+(define-syntax-class fusable-stream-consumer
+  (pattern (~or foldr:fsc-foldr
+                foldl:fsc-foldl
+                list-ref:fsc-list-ref
+                lenght:fsc-length
+                empty?:fsc-empty?
+                default:fsc-default
+                )))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The actual fusion generator implementation
+
+;; Used only in deforest-rewrite to properly recognize the end of
+;; fusable sequence.
+(define-syntax-class non-fusable
+  (pattern (~not (~or _:fusable-stream-transformer
+                      _:fusable-stream-producer
+                      _:fusable-stream-consumer))))
+
+(define (make-deforest-rewrite generate-fused-operation)
+  (lambda (stx)
+    (syntax-parse stx
```

Comment:
Not everybody likes this shorthand, but `(λ (stx) (syntax-parse stx` is
equivalent to `(syntax-parser`; you can use `this-syntax` to access what would
be `stx` under `syntax-parser`. (I also note you wrote `lambda` and not `λ`;
I'll leave it to Sid to prefer one over the other.)

Reply:
Will do in a next PR against the integration branch.